### PR TITLE
Refactor/loading spinner show hide

### PR DESCRIPTION
**Pull request by imnotteixeira**
## Description
Solves videojs/video.js#5056 

## Changes
Created show() and hide() methods to LoadingSpinner class, and changed player.js methods to utilize those loading spinner methods instead of directly changing the DOM.
Also created unit tests for this refactor, which are currently passing.

The show() method adds the `vjs-waiting` css class while the hide() method removes it.
The `vjs-seeking` class was removed, as it was not necessary, because it did the exact same as the `vjs-waiting`

## Requirements Checklist
- [x] Refactor implemented
- [x] Change has been verified in an actual browser (Chome)
- [x] Unit Tests updated or fixed

- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

**Comment by imnotteixeira** [comment]
@ldayananda @gkatsev please review :)

**Comment by gkatsev** [comment]
Unfortunately, the removal of vjs-seeking classes makes this change not backwards compatible.

**Comment by imnotteixeira** [comment]
Ok then, so if I add and remove always the `vjs-seeking` and `vjs-waiting` classes, would that work?

**Comment by gkatsev** [comment]
Maybe, the other problem is the semantic meaning of the change in behavior. Though, it's possible that there isn't a better way of doing it and still staying backwards compatible.

**Comment by mister-ben** [comment]
This should check whether the player has a loading spinner child before calling methods on it. 

I'm concerned that adding both the seeking and the waiting classes in either state would also be a breaking change. I'm sure I've seen customised players with distinct styles for those classes.

**Comment by imnotteixeira** [comment]
Also added the verification of spinner existence as @mister-ben suggested

File: src/js/player.js [hunk 1]
```diff
@@ -1286,6 +1286,11 @@ class Player extends Component {
 
     this.removeClass('vjs-ended');
     this.removeClass('vjs-seeking');
+    const loadingSpinner = this.getChild('loadingSpinner');
```
Comment (BrandonOCasey) [inline src/js/player.js]: I think it would be better to make a `showSpinner` and `hideSpinner` function and call that it each of these places.

Comment (gkatsev) [reply src/js/player.js]: components are already supposed to have show/hide methods, so, I think this is fine

Comment (imnotteixeira) [reply src/js/player.js]: @BrandonOCasey I'm trying not to make the player class feature envious, which is a code smell (https://refactoring.guru/smells/feature-envy), thus the override of existing show and hide methods is appropriate

**Comment by dcm1984** [comment]
Working as intended and the tests should cover all the use cases

**Comment by imnotteixeira** [comment]
@gkatsev Do you think the PR would be eligible to be merged, or would you recommend any more changes?

**Comment by xRuiAlves** [comment]
@gkatsev  Would this PR be ready to be accepted? We (me, @imnotteixeira  and @reeckset) are currently working on this open source project in the context of a university course. 
In this course we are challenged to participate in an open source project, solving two issues and documenting the project's architecture.
It would be very beneficial if this PR could be approved (should not even need to be merged, just have its changes accepted). Thanks in advance!

**Comment by gkatsev** [comment]
You probably need to update your fork with the latest stuff in the master branch.

**Comment by imnotteixeira** [comment]
@gkatsev 

I did that, I also tried just rewriting the package-lock with the one in the master branch, nothing works..

**Comment by xRuiAlves** [comment]
Good afternoon @gkatsev , @mister-ben , @BrandonOCasey ,
As mentioned by my collegue @miguelpduarte in PR #5663, me and an a few more collegues are currently working on videojs in the context of an university course project. The project's due date is tomorrow morning and it would help us a real lot if this PR (aswell as our other PR, #5684) was approved (it needs not even need to be merged; just having it accepted would already be a great help for us!). 
Do you think that that would be possible? Thanks in advance!

File: src/js/loading-spinner.js [hunk 2]
```diff
@@ -34,6 +34,22 @@ class LoadingSpinner extends Component {
 
     return el;
   }
+
+  /**
+   * Show the `LoadingSpinner`s DOM element, by adding the `show` css class to it.
+   */
+  show() {
+    super.show();
+    this.el().classList.add('show');
```
Comment (gkatsev) [inline src/js/loading-spinner.js]: Component has an API for adding and removing classes [`this.addClass`](https://github.com/videojs/video.js/blob/master/src/js/component.js#L721)

File: src/css/components/_loading.scss [hunk 3]
```diff
@@ -20,8 +20,7 @@
   visibility: hidden;
 }
 
-.vjs-seeking .vjs-loading-spinner,
-.vjs-waiting .vjs-loading-spinner {
+.vjs-loading-spinner.show {
```
Comment (gkatsev) [inline src/css/components/_loading.scss]: the `.show` class is too generic. We generally prefix everything with `vjs`

**Comment by gkatsev** [comment]
I'm not sure that we'll be able to approve it in time for your deadline. This change is potentially not backwards compatible and so we'd want to verify that before approving. We won't be able to accept it if it isn't backwards compatible or we're not confidant enough that the change is backwards compatible. I'm unsure whether I'll have enough time to review this PR today and think about its implications in the project.

**Comment by xRuiAlves** [comment]
Allright, we are going to add the requested changes to the PR anyhow. Thank you for the input

**Comment by gkatsev** [comment]
Thanks! We'll definitely spend the time looking at the PR, it just may not happen tonight.

Is there a particular reason why it needs to be approved?

**Comment by xRuiAlves** [comment]
In our course's project, we are "challenged" to take part and collaborate in an open source project of our interest. As I had already worked with video.js quite a lot in my summer internship in the past summer, we ended up trying to collaborate to it! We got bonus points in our final grade assessment if we get our PR accepted by a project's member. Sorry if I was a bit to insistent :sweat_smile:

File: src/js/loading-spinner.js (hunk 2 above)
Comment (imnotteixeira) [reply src/js/loading-spinner.js]: Changed in latest PR

File: src/css/components/_loading.scss (hunk 3 above)
Comment (imnotteixeira) [reply src/css/components/_loading.scss]: Changed in latest commit to .vjs-loading-spinner-show

**Comment by dcm1984** [comment]
I think the Loading Spinner show/hide API actually solves another (older) issue regarding the loading spinner. I can't seem to find it tho, maybe it was marked as stale?

**Comment by gkatsev** [comment]
Finally got a chance to take a look at this in detail. I think the changes look good but as they stand this is *not* backwards compatible. For better or worse, the loading spinner needs to come up when `vjs-waiting` class is added to the player. I wonder if keeping the old CSS around with the rest of the changes will satisfy all requirements.

**Pull request closed without merging**